hypothesis-python/docs/reference/developer.rst

```diff
@@ -0,0 +1,29 @@
+Developer Reference
+===================
+
+.. warning::
+
+    This page documents APIs that Hypothesis makes available to developers building tools, libraries, or research on top of Hypothesis. It is not intended for users of Hypothesis.
+
+    While these APIs are mostly stable, they are considered internal, and are not subject to our standard :ref:`deprecation policy <deprecation-policy>`. We may make breaking changes to these APIs in minor or patch releases.
+
+Alternative backends
+--------------------
+
+.. autoclass:: hypothesis.internal.conjecture.providers.PrimitiveProvider
+    :members:
+    :undoc-members:
```

Comment:
skim through the rendered docs and fix up formatting and add cross-references (e.g. to observability).

I think we should probably write some short docs for each `draw_*` method, and drop `undoc_members` - e.g. that p=0 or p=1 should force a boolean draw, and documenting `IntervalSet` since that's an argument to `draw_string`.

Reply:
I went with numpy style docstrings and auto-type-hints here

Reply:
This was `sphinx_autodoc_typehints` at fault, and nope, getting it to work properly is a pain. I don't want to have to duplicate return types everywhere (I'd like to use numpy-style docstrings for strategies as well), but I think `sphinx_autodoc_typehints` just makes some unfortunate design decisions for us. 

I bet we can roll our own simple variant in the future. I'm dropping `sphinx_autodoc_typehints` for now and just manually specifying the return type.